File: library-frontend/src/services/LoanService.js
import axios from 'axios'

const LOANS_URL = 'http://localhost:8080/loans';
let response;

class LoanService {

  getByUserId(userId) {
    const response = axios({
      method: 'get',
      url: LOANS_URL + "/user/" + userId,
      headers: { 
        'Authorization': `Bearer ${localStorage.getItem('user-token')}`
      }
    })
    .catch(function (error) {
      console.log(error);
    });

    return response;
  }

  create(userId, editionId) {
    const response = axios({
      method: 'post',
      url: LOANS_URL,
      params: {
        userId: userId,
        editionId: editionId
      },
      headers: { 
        'Authorization': `Bearer ${localStorage.getItem('user-token')}`
      },
    })
    .catch(function (error) {
      console.log(error);
    });

    return response;
  }

  update(loan) {
    const response = axios({
      method: 'put',
      url: LOANS_URL,
      data: loan,
      headers: { 
        'Authorization': `Bearer ${localStorage.getItem('user-token')}`
      },
    })
    .catch(function (error) {
      console.log(error);
    });

    return response;
  }
  
  // set loan status to finalized and its returnedOn date
  async finalize(loanId) {
    try {
      response = await axios.put(LOANS_URL + '/' + loanId + '/finalize', {}, {
        headers: {
          Authorization: `Bearer ${localStorage.getItem('user-token')}`
        },
      });
    } catch (error) {
      console.log(error.message);
    }
    return response;
  }

  // set loan status to active and its loanedOn date
  async activate(loanId) {
    console.log(loanId);
    try {
      response = await axios.put(LOANS_URL + "/" + loanId + '/activate', {}, {
        headers: {
          Authorization: `Bearer ${localStorage.getItem('user-token')}`
        },
      });
    } catch (error) {
      console.log(error.message);
    }
    return response;
  }

  // set loan status to canceled 
  async cancel(loanId) {
    try {
      response = await axios.put(LOANS_URL + "/" + loanId + '/cancel', {}, {
        headers: {
          Authorization: `Bearer ${localStorage.getItem('user-token')}`
        },
      });
    } catch (error) {
      console.log(error.message);
    }
    return response;
  }

  async setChargeAsPaid(loanId) {
    try {
      response = await axios.put(LOANS_URL + "/" + loanId + '/chargePaid', {}, {
        headers: {
          Authorization: `Bearer ${localStorage.getItem('user-token')}`
        },
      });
    } catch (error) {
      console.log(error.message);
    }
    return response;
  }
}

export default new LoanService();

File: library-backend/src/main/java/com/adrian/library/loan/LoanService.java
package com.adrian.library.loan;

import com.adrian.library.reservation.BookAlreadyReservedException;
import com.adrian.library.reservation.Reservation;
import com.adrian.library.user.UserLacksActionPointsException;

import java.util.List;

public interface LoanService {
    void init();
    List<Loan> getAll();
    List<Loan> getByUserId(Integer id);
    Loan get(Integer id);
    Loan create(Integer userId, Integer editionId, Reservation reservation)
            throws BookAlreadyLoanedException, BookAlreadyReservedException, UserLacksActionPointsException;
    Loan update(Loan loan);
    void finalize(Integer id) throws BookAlreadyLoanedException, BookAlreadyReservedException, UserLacksActionPointsException;
    void activate(Integer id);
    void setChargeAsPaid(Integer id);
    void cancel(Integer id) throws BookAlreadyReservedException, BookAlreadyLoanedException, UserLacksActionPointsException;
    void cancelLoans() throws UserLacksActionPointsException, BookAlreadyReservedException, BookAlreadyLoanedException;
    void chargeLoans();
}
